llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp

```diff
@@ -14915,6 +14920,181 @@ SDValue DAGCombiner::visitEXTEND_VECTOR_INREG(SDNode *N) {
+         "Unexpected types for truncate operation");
+
+  APInt Max;
+  if (SDValue UMin = matchMinMax(In, ISD::UMIN, Max, /*Signed*/ false)) {
+    // C2 should be equal to UINT32_MAX / UINT16_MAX / UINT8_MAX according
+    // the element size of the destination type.
+    if (Max.isMask(VT.getScalarSizeInBits()))
+      return UMin;
+  }
+
+  return SDValue();
+}
+
+/// Detect patterns of truncation with signed saturation:
+/// (truncate (smin ((smax (x, signed_min_of_dest_type)),
+///                  signed_max_of_dest_type)) to dest_type)
+/// or:
+/// (truncate (smax ((smin (x, signed_max_of_dest_type)),
+///                  signed_min_of_dest_type)) to dest_type).
+/// With MatchPackUS, the smax/smin range is [0, unsigned_max_of_dest_type].
+/// Return the source value to be truncated or SDValue() if the pattern was not
+/// matched.
+static SDValue detectSSatSPattern(SDValue In, EVT VT) {
+  unsigned NumDstBits = VT.getScalarSizeInBits();
+  unsigned NumSrcBits = In.getScalarValueSizeInBits();
+  assert(NumSrcBits > NumDstBits && "Unexpected types for truncate operation");
+
+  APInt SignedMax, SignedMin;
+  SignedMax = APInt::getSignedMaxValue(NumDstBits).sext(NumSrcBits);
+  SignedMin = APInt::getSignedMinValue(NumDstBits).sext(NumSrcBits);
+  if (SDValue SMin = matchMinMax(In, ISD::SMIN, SignedMax, /*Signed*/ true)) {
+    if (SDValue SMax =
+            matchMinMax(SMin, ISD::SMAX, SignedMin, /*Signed*/ true)) {
+      return SMax;
+    }
+  }
+  if (SDValue SMax = matchMinMax(In, ISD::SMAX, SignedMin, /*Signed*/ true)) {
+    if (SDValue SMin =
+            matchMinMax(SMax, ISD::SMIN, SignedMax, /*Signed*/ true)) {
+      return SMin;
+    }
+  }
+  return SDValue();
+}
+
+/// Detect patterns of truncation with unsigned saturation:
+static SDValue detectSSatUPattern(SDValue In, EVT VT, SelectionDAG &DAG,
+                                  const SDLoc &DL) {
+  EVT InVT = In.getValueType();
+
+  // Saturation with truncation. We truncate from InVT to VT.
+  assert(InVT.getScalarSizeInBits() > VT.getScalarSizeInBits() &&
+         "Unexpected types for truncate operation");
+
+  APInt Min, Max;
+  SDValue SMax, SMin, UMin, First;
+  // (truncate (smin (smax (x, Min), Max)) to dest_type),
+  // (truncate (smax (smin (x, Max), Min)) to dest_type)
+  // (truncate (umin (smax (x, Min), Max)) to dest_type)
+  // where Min >= 0, Max is unsigned max of destination type and Min <= Max.
+  if (First = SMax = matchMinMax(In, ISD::SMAX, Min, /*Signed*/ false))
+    SMin = matchMinMax(First, ISD::SMIN, Max, /*Signed*/ false);
+  else if (First = SMin = matchMinMax(In, ISD::SMIN, Max, /*Signed*/ false))
+    SMax = matchMinMax(First, ISD::SMAX, Min, /*Signed*/ false);
+  else if (First = UMin = matchMinMax(In, ISD::UMIN, Max, /*Signed*/ false))
+    SMax = matchMinMax(UMin, ISD::SMAX, Min, /*Signed*/ false);
+
+  if (SMax && Min.isNonNegative() && Max.isMask(VT.getScalarSizeInBits())) {
```

Comment:
Can you explain this Min.isNonNegative() bit, and why it is needed?

Reply:
This is because we want to check the range of unsigned values. The minimum value must be greater than or equal to zero.

Reply:
I feel that the non-negative part is complicating things. Can we keep this patch simpler and just match the exact pattern. Something like:
```
  APInt UnsignedMax, Zero;
  UnsignedMax = APInt::getMaxValue(NumDstBits).zext(NumSrcBits);
  Zero = APInt::getZero(NumSrcBits);

  // (truncate (smin (smax (x, 0), Max)) to dest_type),
  if (Max = matchMinMax(In, ISD::SMAX, Zero)) {
    if (Min = matchMinMax(Max, ISD::SMIN, UnsignedMax))
      return Min;
    if (Min = matchMinMax(Max, ISD::UMIN, UnsignedMax))
      return Min;
  }
  if (Min = matchMinMax(In, ISD::SMIN, UnsignedMax))
    if (Max = matchMinMax(Min, ISD::SMAX, Zero))
      return Max;
  if (Min = matchMinMax(In, ISD::UMIN, UnsignedMax))
    if (Max = matchMinMax(Min, ISD::SMAX, Zero))
      return Max;

  return SDValue();
```

We can add the extra bits in a follow-up if they are needed.

Reply:
how about now? do you think I'm doing properly?

Reply:
we need to considering when Min > 0. so, we need to get the Min value from matchMinMax.

Reply:
we need to considering when Min > 0. 
so, &C was not removed to get the Min value from matchMinMax.

Reply:
@davemgreen how about this? 
```
static SDValue matchMin(SDValue V, unsigned Opcode, APInt &Limit) {
  if (V.getOpcode() == Opcode) {
    APInt C;
    if (ISD::isConstantSplatVector(V.getOperand(1).getNode(), C))
      if (C == Limit)
        return V.getOperand(0);
      if (C > Limit && Limit == 0)
        return V.getOperand(0);
  }
  return SDValue();
}
```

Reply:
```
  // Min == 0, Max is unsigned max of destination type.
  if (SMax = matchMinMax(In, ISD::SMAX, Zero)) {
    if (SMin = matchMinMax(In, ISD::SMIN, UnsignedMax))
      return SMin.getOperand(0);
  }
  // FIX THIS, Min > 0, Max is unsigned max of destination type and Min <= Max.
  if (SMax = matchMinMax(In, ISD::SMAX, OVER_ZERO)) {
    if (SMin = matchMinMax(In, ISD::SMIN, UnsignedMax))
      return DAG.getNode(ISD::SMAX, DL, InVT, SMin, In.getOperand(1));
  }
  // Max is unsigned max of destination type.
  if (SMin = matchMinMax(In, ISD::SMIN, UnsignedMax)) {
    // Min == 0
    if (SMax = matchMinMax(In, ISD::SMAX, Zero))
      return SMax.getOperand(0);
    // FIX THIS, Min > 0, Max is unsigned max of destination type and Min <= Max.
    if (SMax = matchMinMax(In, ISD::SMAX, OVER_ZERO))
      return SMin;
  }
  if (UMin = matchMinMax(In, ISD::UMIN, UnsignedMax)) {
    // Min == 0
    if (SMax = matchMinMax(In, ISD::SMAX, Zero))
      return SMax.getOperand(0);
    // FIX THIS, Min > 0, Max is unsigned max of destination type and Min <= Max.
    if (SMax = matchMinMax(In, ISD::SMAX, OVER_ZERO))
      return UMin;
  }
```
I made a temporary code of what it would look like if I changed to the logic you showed in your example, where I want to modify the part that says `OVER_ZERO` to check if `Min>0`. What do you think?
@davemgreen

Reply:
```
// Match min and return limit value as a parameter.
static SDValue matchMinMax(SDValue V, unsigned Opcode, APInt &Limit) {
  if (V.getOpcode() == Opcode) {
    APInt C;
    if (ISD::isConstantSplatVector(V.getOperand(1).getNode(), C)) {
      if (C == Limit)
        return V.getOperand(0);
      if (V.getOpcode() == ISD::SMAX && !Limit.isNegative() && C.sgt(Limit) && Limit != 0)
        return V.getOperand(0);
    }
  }
  return SDValue();
}
```
I've solved the above homework by excluding zero cases from `matchMinMax`, what do you think?

Reply:
```

  // Min == 0, Max is unsigned max of destination type.
  if (SMax = matchMinMax(In, ISD::SMAX, Zero)) {
    if (SMin = matchMinMax(SMax, ISD::SMIN, UnsignedMax))
      return SMax.getOperand(0);
  }
  // Min >= 1, Max is unsigned max of destination type and Min <= Max.
  if (SMax = matchMinMax(In, ISD::SMAX, Zero)) {
    if (SMin = matchMinMax(SMax, ISD::SMIN, UnsignedMax))
      return DAG.getNode(ISD::SMAX, DL, InVT, SMin, In.getOperand(1));
  }
  // Max is unsigned max of destination type.
  if (SMin = matchMinMax(In, ISD::SMIN, UnsignedMax)) {
    // Min == 0
    if (SMax = matchMinMax(SMin, ISD::SMAX, Zero))
      return SMin.getOperand(0);
    // Min >= 1, Max is unsigned max of destination type and Min <= Max.
    if (SMax = matchMinMax(SMin, ISD::SMAX, Zero))
      return SMin;
  }
  if (UMin = matchMinMax(In, ISD::UMIN, UnsignedMax)) {
    // Min == 0
    if (SMax = matchMinMax(UMin, ISD::SMAX, Zero))
      return UMin.getOperand(0);
    // Min >= 1, Max is unsigned max of destination type and Min <= Max.
    if (SMax = matchMinMax(UMin, ISD::SMAX, Zero))
      return;
```
here is full code

Reply:
```

// Match min and return limit value as a parameter.
static SDValue matchMinMax(SDValue V, unsigned Opcode, APInt &Limit) {
  if (V.getOpcode() == Opcode) {
    APInt C;
    if (ISD::isConstantSplatVector(V.getOperand(1).getNode(), C)) {
      if (C == Limit)
        return V.getOperand(0);
      if (V.getOpcode() == ISD::SMAX && !Limit.isNegative() && C.uge(Limit) &&
          Limit != 0)
        return V.getOperand(0);
    }
  }
  return SDValue();
}
```
```

  APInt One = APInt(NumSrcBits, 1);
  
  // Min == 0, Max is unsigned max of destination type.
  if (SMax = matchMinMax(In, ISD::SMAX, Zero)) {
    if (SMin = matchMinMax(SMax, ISD::SMIN, UnsignedMax))
      return SMax.getOperand(0);
  }
  // Min >= 1, Max is unsigned max of destination type and Min <= Max.
  if (SMax = matchMinMax(In, ISD::SMAX, One)) {
    if (SMin = matchMinMax(SMax, ISD::SMIN, UnsignedMax))
      return DAG.getNode(ISD::SMAX, DL, InVT, SMin, In.getOperand(1));
  }
  // Max is unsigned max of destination type.
  if (SMin = matchMinMax(In, ISD::SMIN, UnsignedMax)) {
    // Min == 0
    if (SMax = matchMinMax(SMin, ISD::SMAX, Zero))
      return SMin.getOperand(0);
    // Min >= 1
    if (SMax = matchMinMax(SMin, ISD::SMAX, One))
      return SMin;
  }
  // Max is unsigned max of destination type
  if (UMin = matchMinMax(In, ISD::UMIN, UnsignedMax)) {
    // Min == 0
    if (SMax = matchMinMax(UMin, ISD::SMAX, Zero))
      return UMin.getOperand(0);
    if (SMax = matchMinMax(UMin, ISD::SMAX, One))
      return UMin;
  }
  ```
I've solved the above homework by excluding zero cases from `matchMinMax`, what do you think?

Reply:
Can we just keep the basic version in this first patch? The patterns we want to match are:
USAT_U: `trunc(umin(x, 255))`
SSAT_S: `trunc(smin(smax(x, -128), 127))`
SSAT_U: `trunc(smin(smax(x, 0), 255))`

The min/max could be in the opposite orders, and the last one is odd because it can match `trunc(umin(smax(x, 0), 255))` as the umin and smin are equivalent.
I don't believe we should need any >= checks at the moment, and if we do they need tests. IMO they would be better adding in a follow-up patch, so we can get the basics down first. They might well enable the creating of more saturating nodes, but with this patch adding all three variants there is already quite a lot going on.